google-cloud-spanner/src/main/java/com/google/cloud/spanner/OpenTelemetryApiTracer.java

```diff
@@ -0,0 +1,276 @@
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.threeten.bp.Duration;
+
+/**
+ * {@link com.google.api.gax.tracing.ApiTracer} for use with OpenTelemetry. Based on {@link
+ * com.google.api.gax.tracing.OpencensusTracer}.
+ */
+class OpenTelemetryApiTracer implements ApiTracer {
+  private final AttributeKey<Long> ATTEMPT_COUNT_KEY = AttributeKey.longKey("attempt.count");
+  private final AttributeKey<Long> TOTAL_REQUEST_COUNT_KEY =
+      AttributeKey.longKey("total_request_count");
+  private final AttributeKey<Long> TOTAL_RESPONSE_COUNT_KEY =
+      AttributeKey.longKey("total_response_count");
+  private final AttributeKey<String> EXCEPTION_MESSAGE_KEY =
+      AttributeKey.stringKey("exception.message");
+  private final AttributeKey<Long> ATTEMPT_NUMBER_KEY = AttributeKey.longKey("attempt.number");
+  private final AttributeKey<Long> ATTEMPT_REQUEST_COUNT_KEY =
+      AttributeKey.longKey("attempt.request_count");
+  private final AttributeKey<Long> ATTEMPT_RESPONSE_COUNT_KEY =
+      AttributeKey.longKey("attempt.response_count");
+  private final AttributeKey<String> CONNECTION_ID_KEY = AttributeKey.stringKey("connection");
+  private final AttributeKey<Long> RETRY_DELAY_KEY = AttributeKey.longKey("delay_ms");
+  private static final AttributeKey<Long> BATCH_SIZE_KEY = AttributeKey.longKey("batch.size");
+  private static final AttributeKey<Long> BATCH_COUNT_KEY = AttributeKey.longKey("batch.count");
+
+  private final Span span;
+  private final OperationType operationType;
+
+  private volatile String lastConnectionId;
+  private volatile long currentAttemptId;
+  private final AtomicLong attemptSentMessages = new AtomicLong(0);
+  private long attemptReceivedMessages = 0;
+  private final AtomicLong totalSentMessages = new AtomicLong(0);
+  private long totalReceivedMessages = 0;
+
+  OpenTelemetryApiTracer(@Nonnull Span span, @Nonnull OperationType operationType) {
+    this.span = Preconditions.checkNotNull(span);
+    this.operationType = Preconditions.checkNotNull(operationType);
+  }
+
+  Span getSpan() {
+    return this.span;
+  }
+
+  @Override
+  public Scope inScope() {
+    final io.opentelemetry.context.Scope openTelemetryScope = span.makeCurrent();
+    return openTelemetryScope::close;
+  }
+
+  @Override
+  public void operationSucceeded() {
+    span.setAllAttributes(baseOperationAttributes());
+    span.setStatus(StatusCode.OK);
+    span.end();
+  }
+
+  @Override
+  public void operationCancelled() {
+    span.setAllAttributes(baseOperationAttributes());
+    span.setStatus(StatusCode.ERROR, "Cancelled by caller");
+    span.end();
+  }
+
+  @Override
+  public void operationFailed(Throwable error) {
+    span.setAllAttributes(baseOperationAttributes());
+    span.setStatus(StatusCode.ERROR, error.getMessage());
+    span.end();
+  }
+
+  @Override
+  public void lroStartFailed(Throwable error) {
```

Comment:
It's called from here: https://github.com/googleapis/sdk-platform-java/blob/5799827a864be627bac03969cc178efc9d6170aa/gax-java/gax/src/main/java/com/google/api/gax/tracing/TracedOperationInitialCallable.java#L85

This test case covers the kind of failure that would trigger this method to be called: https://github.com/googleapis/java-spanner/blob/bfa1aac17f69ce2e365c24133a5eb5cef205f81e/google-cloud-spanner/src/test/java/com/google/cloud/spanner/OpenTelemetryApiTracerTest.java#L346

I agree that it is relatively unlikely to happen in production, as it requires an RPC that would normally return an LRO to fail in a way that does not return an LRO, but instead just returns an error. But it is needed to implement the interface, and in theory it could happen (I could for example imagine an RPC like this returning a RESOURCE_EXHAUSTED error directly, instead of an LRO, if there is a limit on the number of operations running in parallel).

Reply:
> it is needed to implement the interface

All methods in `ApiTracer` are default no-op methods now, you don't have to override them. 

What I'm trying to get to is, maybe we can start with traces/events that we know will be most useful, we don't have to implement each and every scenario. Especially for LROs and batches related methods, they may not be well tested before and I'm not confident enough that they are working as intended. 
For example, maybe `lroStartFailed ` should be called in places more than just in `TracedOperationInitialCallable`, and we did find a few small bugs related to the whole `ApiTracer` framework while implementing OpenTelemetry metrics in Gax. 

That being said, if you do think they provide value to Spanner, we can start with it and fix bugs along the way when we find them.

Reply:
I would prefer to keep this, because:
1. Tracing long-running operations adds value for Spanner, specifically because DDL statements are executed as long-running operations. Giving customers (and us) insights into this can help in debugging issues, especially if the customer uses statements like `create table if not exists...`, which are relatively quick to execute, but still add significant latency if they are executed as part of for example the standard application startup.
2. This method covers one of the potential end states of such an LRO. Failing to implement that method, would leave the trace open.
3. If there are specific cases that do not call this method, then that will also leave the trace open, but getting the trace closed in most cases is still better than not implementing this method. The 'standard' way that this method should be called is covered by a test case, so I feel confident that the most common code flow that should be covered by this method works as intended.